aws-datastore/src/main/java/com/amplifyframework/datastore/AWSDataStorePlugin.java

```diff
@@ -505,6 +505,24 @@ private void beforeOperation(@NonNull final Runnable runnable) {
         }
     }
 
+    private void beforeOperationReadyStorage(@NonNull final Runnable runnable) {
+        try {
+            Completable.fromAction(
+                () -> {
+                    categoryInitializationsPending.await();
+                    orchestrator.startWithBehaviorExitImmediatelyIfTransitioning();
+                })
+                .andThen(Completable.fromRunnable(runnable))
+                .blockingAwait(LIFECYCLE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+        } catch (Throwable throwable) {
+            if (!orchestrator.isStarted()) {
+                LOG.warn("Failed to execute request because DataStore is not fully initialized.");
+            } else {
+                LOG.warn("Failed to execute request due to an unexpected error.", throwable);
+            }
+        }
+    }
+
```

Comment:
Maybe instead of having two `beforeOperation*` and two `start*` methods, you could use an enum to define possible strategies for starting the orchestrator.

For example, you could create an enum inside the Orchestrator class called `StartWaitStrategy` with two items: `StorageAndSyncReady` and `StorageReady`.

Then in the `AWSDataStorePlugin` class, instead of having two `beforeOperation*` methods, you could have one with `Orchestrator.StartWaitStrategy` as a parameter: 

```java
private void beforeOperation(Orchestrator.StartWaitStrategy strategy, @NonNull final Runnable runnable) {
   ...
      orchestrator.start(strategy);
   ...
}
```
Likewise, you could apply a similar pattern to the start method in the Orchestrator. 

```java
public void start(StartWaitStrategy strategy) {
   if(StartWaitStrategy.StorageReady.equals(strategy)) {
     attemptStart(NO_WAIT_TIMEOUT);
   } else {
     attemptStart(LOCAL_OP_TIMEOUT_SECONDS);
   }
}
```

Let me know what you think.

Reply:
I like your suggestion @rjuliano! What do you think, @saltonmassally?

Reply:
Sounds solid; I will make the changes